rfcs/text/0004_application_service_mounting.md

```diff
@@ -0,0 +1,238 @@
+- Start Date: 2019-05-10
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A front-end service to manage registration and root-level routing for
+first-class applications.
+
+# Basic example
+
+```tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+class MyPlugin {
+  setup({ application }) {
+    application.register({
+      id: 'my-app',
+      title: 'My Application',
+      async mount(context, targetDomElement) {
+        const { MyApp } = await import('./components');
+
+        ReactDOM.render(
+          <MyApp mountContext={context} deps={pluginStart} />,
+          targetDomElement
+        );
+
+        return () => {
+          ReactDOM.unmountComponentAtNode(targetDomElement);
+        };
+      }
+    });
+  }
+}
+```
+
+# Motivation
+
+By having centralized management of applications we can have a true single page
+application. It also gives us a single place to enforce authorization and/or
+licensing constraints on application access.
+
+By making the mounting interface of the ApplicationService generic, we can
+support many different rendering technologies simultaneously to avoid framework
+lock-in.
+
+# Detailed design
+
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with Kibana to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+## Interface
+
+```ts
+/** A context type that implements the Handler Context pattern from RFC-0003 */
+export interface MountContext {
+  core: {
+    chrome: ChromeStart;
+    http: HttpStart;
+    i18n: I18nStart;
+    notifications: NotificationStart;
+    overlays: OverlayStart;
+    uiSettings: UISettingsStart;
+  }
+  // Other plugins can inject context by registering additional context providers
+  [contextName: string]: unknown;
+}
+
+export type Unmount = () => Promise<void> | void;
+
+export interface AppSpec {
+  /**
+   * A unique identifier for this application. Used to build the route for this
+   * application in the browser.
+   */
+  id: string;
+
+  /**
+   * The title of the application.
+   */
+  title: string;
+
+  /**
+   * A mount function called when the user navigates to this app's `rootRoute`.
```

Comment:
I think `rootRoute` was removed in a previous edit, so this reference isn't clear